File: data_structures/sqrt_root_decomposition/Python/sqrt_decomposition.py
import math

block = [0]*100
arr = [0]*10000

def preprocess(seq, n):
	
	# Initializing block pointer and claculating block size
	ind = -1
	block_size = math.sqrt(n)

	# calculating the value of each block : 
	for i in range(n):
		
		arr[i]= seq[i]
		if i%block_size == 0:
			ind+=1

		block[ind] += arr[i]

	return block_size

def query(l,r):

	res=0

	# Traversing first block :
	while l<r and l%block_size!=0 and l!=0 :
		res+=arr[l]
		l+=1

	# Travering all middle blocks :
	while l+block_size <= r :
		res += block[l//block_size]
		l+=block_size

	# Travering the last block :
	while l<=r :
		res+=arr[l]
		l+=1

	return res

def update(ind, val):

	# Updating the value of each block and array :
	block_ind = ind//block_size
	block[block_ind] += val - arr[ind]
	arr[ind]=val


''' Test Program Starts Here '''

seq = [1, 5, 2, 4, 6, 1, 3, 5, 7, 10];
n=len(seq)

block_size = preprocess(seq, n)

print("Query(3,8) : ", query(3,8))
print("Query(1,6) : ", query(1,6))
print("Query(8,8) : ", query(8,8))

File: data_structures/sqrt_root_decomposition/C++/sqrtdecomposition.cpp
#include "iostream"
#include "math.h"
using namespace std;

#define MAXN 10000
#define SQRSIZE  100

int arr[MAXN];               // original array
int block[SQRSIZE];          // decomposed array
int blk_sz;                      // block size

// Time Complexity : O(1)
void update(int idx, int val)
{
	int blockNumber = idx / blk_sz;
	block[blockNumber] += val - arr[idx];
	arr[idx] = val;
}

// Time Complexity : O(sqrt(n))
int query(int l, int r)
{
	int sum = 0;
	while (l<r and l%blk_sz!=0 and l!=0)
	{
		// traversing first block in range
		sum += arr[l];
		l++;
	}
	while (l+blk_sz <= r)
	{
		// traversing completely overlapped blocks in range
		sum += block[l/blk_sz];
		l += blk_sz;
	}
	while (l<=r)
	{
		// traversing last block in range
		sum += arr[l];
		l++;
	}
	return sum;
}

// Fills values in input[]
void preprocess(int input[], int n)
{
	// initiating block pointer
	int blk_idx = -1;

	// calculating size of block
	blk_sz = sqrt(n);

	// building the decomposed array
	for (int i=0; i<n; i++)
	{
		arr[i] = input[i];
		if (i%blk_sz == 0)
		{
			// entering next block
			// incementing block pointer
			blk_idx++;
		}
		block[blk_idx] += arr[i];
	}
}

// Driver code
int main()
{
	// We have used separate array for input because
	// the purpose of this code is to explain SQRT
	// decomposition in competitive programming where
	// we have multiple inputs.
	int input[] = {1, 5, 2, 4, 6, 1, 3, 5, 7, 10};
	int n = sizeof(input)/sizeof(input[0]);

	preprocess(input, n);

	cout << "query(3,8) : " << query(3, 8) << endl;
	cout << "query(1,6) : " << query(1, 6) << endl;
	update(8, 0);
	cout << "query(8,8) : " << query(8, 8) << endl;
	return 0;
}
